### ISSUE TITLE
Structural search and OR branches

### ISSUE THREAD

**Issue by sqs** [sourcegraph/sourcegraph#14081]
I tried to use the following query:

[`repo:^github\.com/sourcegraph/sourcegraph$ (var :[[name]] interface { :[2] }) or (func :[[1]](:[2] interface { :[3] }))`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+%28var+:%5B%5Bname%5D%5D+interface+%7B+:%5B2%5D+%7D%29+or+%28func+:%5B%5B1%5D%5D%28:%5B2%5D+interface+%7B+:%5B3%5D+%7D%29%29&patternType=structural)

(I was looking for Go inline interfaces, such as `var x interface { Foo() string }` or `func Foo(x interface { Bar() int })`.)

I didn't *expect* it would work, and it didn't. (It finds no results, but if I try each of the or-branches separately, they do have results.) As a user, I have a few questions:

- Why doesn't it work? We should notify the user of this limitation (or just document it in the meantime).
- Are the hole names shared across or-branches? (I.e., is it OK that I reused `:[2]` in the 2nd or-branch?) I don't think there is an obvious answer here. I suppose it could be useful to reuse holes in multiple and-branches. In or-branches, if both branches match, it could also check that the matches for equal hole names are equal (but that seems weird). But none of these behaviors are "obviously the only correct way".

(Low priority.)

/cc @sourcegraph/search-platform

**Comment by rvantonder** [comment]
> Why doesn't it work? We should notify the user of this limitation (or just document it in the meantime).

A leading question to guide this: why are parentheses added around those individual search patterns? I.e., what difference would you expect their presence to make, if the intuition is that they just wrap the pattern and don't functionally do anything?

From an interpretation standpoint, it's ambiguous what the outer parentheses (around the patterns) refer to in that query: should they be searched literally, or is it an expression group in Sourcegraph's language? Because of the structure of the query (more explained below), the parentheses around the structural pattern are interpreted, best-effort, as part of the pattern (search them literally), not an expression group. If we didn't, you could basically not express a lisp-y syntax pattern in structural search, e.g., a lisp pattern like `(...)` would strip away the parens, and we would have to introduce an escape mechanism (or you would have to write out `content:"(lisp :[lisp] lisp thing)"` every time. Tedious). If it weren't obvious to me (human) that you were writing a search pattern for Go, the parens around those patterns may very well have referred to some lisp or s-expression syntax that you'd want to search, so the ambiguity is real. That's the tradeoff that's happening here. The solution to my mind is more obviously communicating how objectively ambiguous syntax is interpreted (disambiguated) by our parser through a combination of syntax highlighting or alerts. My personal take is that we should syntax highlight parentheses that form part of a group expression, and leave parentheses-interpreted-as-patterns unhighlighted as a visual cue.

---

So let's look at getting the result you expected. The original query is parsed as `<repo:thing> <pattern> or <pattern>`. 
The query you wanted to express can be done completely without parentheses, just remove them:
[`repo:^github\.com/sourcegraph/sourcegraph$ var :[[name]] interface { :[2] } or func :[[1]](:[2] interface { :[3] })`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+var+:%5B%5Bname%5D%5D+interface+%7B+:%5B2%5D+%7D+or+func+:%5B%5B1%5D%5D%28:%5B2%5D+interface+%7B+:%5B3%5D+%7D%29&patternType=structural)

Alternatively, if you did want to explicitly group the patterns in an expression, apart from `repo`, a form like `<repo:thing> (<pattern> or <pattern>)` is also totally fine, and equivalent:
[`repo:^github\.com/sourcegraph/sourcegraph$ (var :[[name]] interface { :[2] } or func :[[1]](:[2] interface { :[3] }))`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+%28var+:%5B%5Bname%5D%5D+interface+%7B+:%5B2%5D+%7D+or+func+:%5B%5B1%5D%5D%28:%5B2%5D+interface+%7B+:%5B3%5D+%7D%29%29&patternType=structural)

In the second case, the parser is interprets the outer parens as a group (which is more natural, since it forms an or-expression), and not the pattern. This second case reveals some of how we choose to interpret the parentheses: if balanced parentheses surround an operator like `and` or `or`, we take it to imply grouping/precedence, not something that's part of a pattern syntax. 

TLDR: enclosing a pattern with parentheses, like `(var :[[name]] blah blah)` would normally be redundant, those outer parentheses add no meaning to the query. If we assumed they have no meaning, we lock ourselves out of the option to consider them as part of a syntax pattern (a likely intent that's useful in the context where outer parens refer to syntax). So, instead, we interpret those parentheses literally, as part of the pattern--their presence more usefully suggests an intent to search for them, than the alternative. The working examples above offer two ways to get the result you intended (either no parentheses, or parentheses around an operator), so a natural way to express this certainly exists, hence why I propose a solution using visual cues or nudges, and not changing how the original query is interpreted.

---

> Are the hole names shared across or-branches? 

No, they aren't currently. Think of each pattern creating an environment of matches `var -> value`. These environments are not merged/checked for equality across `and`/`or` expression yet. I'll take care to document this part.

I think the intuitive thing here is likely to merge them (because you can always use new/fresh variable names if you don't want to check for equality across patterns)--being able to check for equality across them is strictly more expressive capability (implies better). But it does complicate evaluating the query because normally we can shortcircuit a series of `or` expressions once we have enough matches. When there's a dependency (equality relation to check across say `or` expressions) we can't do shortcircuiting. 


---

FYI you can now use `...` for better readability of `:[_]`: 
[`repo:^github\.com/sourcegraph/sourcegraph$ var :[[name]] interface {...} or func :[[1]](... interface {...})`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+var+:%5B%5Bname%5D%5D+interface+%7B...%7D+or+func+:%5B%5B1%5D%5D%28...+interface+%7B...%7D%29&patternType=structural)

**Comment by jtibshirani** [comment]
Closing this out, as we are not actively investing in structural search.
